docs/python.md

```diff
@@ -0,0 +1,86 @@
+# Introduction
+
+> [!NOTE]
+> The following is valid as of 2024-03-18. Future versions of `jsii` might provide a more Pythonic approach.
+
+Using the python version of this package requires some extra steps to ensure everything works as expected.
+This document highlights only those steps, so for the basics, please refer to the [main README](../README.md), the [API docs](../API.md) and the [Typescript docs](typescript.md).
+
+## How to implement interfaces
+
+Since this package is written in Typescript and then exported to Python via `jsii`, using interfaces to pass props is not as straightforward as it should be.
+
+Whereas in Typescript you can simply pass the props like this:
+
+```ts
+const costMonitoring = new ApplicationCostMonitoring(
+    monitoringStack,
+    // ⬇︎⬇︎⬇︎ The props ⬇︎⬇︎⬇︎
+    {
+        applicationName: "my-application",
+        monthlyBudget: 200,want)
+        otherStacksIncludedInBudget: [secondStack, firstStack],
+        subscribers: ["alert@example.com"],
+    }
+);
+```
+
+In python it is not so simple.
+
+The correct, and ONLY working approach is to:
+
+1. Create a new class (no inheritance)
+2. Decorate it with `@jsii.implements(<IMyInterface>)`
```

Comment:
This is a weird requirement. The whole CDK stack in Python runs on `jsii`, and we don't need to pass anything to it using this method. Can you please share related resources here?

Reply:
If needed, we should abstract this away from the final consumer.

Reply:
@shahinism 
related (re)-sources (also included in the docs):
- <https://aws.github.io/jsii/user-guides/lib-user/language-specific/python/>
- <https://github.com/aws/jsii/issues/1027>

> we should abstract this away from the final consumer.
I don't think this is possible in a pure Typescript implementation, unless you want to create a dedicated Python one

Reply:
I am not sure about the Python usage, but like @shahinism mentions. When you are using the construct you should not be concerned about anything JSII, it should just be normal Python usage. In an ideal world. So I don't know about this one.

Reply:
I might be wrong but is your Python example in the doc using the construct or inheriting from it? Maybe that edge case only exists if you inherit in Python?

Reply:
> is your Python example in the doc using the construct or inheriting from it?

@rehanvdm 
Are you referring specifically to the interface `IApplicationCostMonitoring`? In that case, neither.

Inheritance doesn't work, nor does it work to use the "interface" directly:

1. in the first case, inheritance, the program will crash with error: `TypeError: Don't know how to convert object to JSON`
2. in the second case, using the construct directly, you can't because the program considers it an abstract class.

All possible "pythonic" approaches (that don't work) have been clearly documented by @Soheil96 [here](https://brenntag-global.atlassian.net/browse/ADAP-602?focusedCommentId=81469).

You can try it yourself: https://git.digib.io/inventory/iberia/salesforce-data-products-glue/-/merge_requests/504.

The CDK run will fail nonetheless because lack of permissions, but the only case in which it fails because of a permission error rather than some type error or other programmatic issues is the one documented here.

Reply:
>     * [python: implementing interface property requires a `@property` declaration aws/jsii#1027](https://github.com/aws/jsii/issues/1027)
The concern in this link is implementing the interface directly in Python instead of using one in TS. Am I wrong?

Reply:
I also spent some time this morning reviewing some of the constructs available at https://constructs.dev/
We can check any supporting Python construct without taking extra care with JSII. Let's see what they are doing differently at the library level.

Reply:
I've spent some time going through the cdk code as suggested. Here's my findings (in no particular order):

- Constructs that have Interfaces in the constructor, when translated to python, expect an object that matches that interface. A good example to follow is the `IConnection` interface and related `Connection` class ([here](https://constructs.dev/packages/@aws-cdk/aws-glue-alpha/v/2.133.0-alpha.0/api/Connection?lang=python) and [here](https://github.com/aws/aws-cdk/blob/4ff3565a9d7b0298bf884822fecabdd3cff643aa/packages/%40aws-cdk/aws-glue-alpha/lib/connection.ts#L124)).
- Notice how Connection has `subnet` and `security_groups` args, but it expects proper `ec2.Subnet()` and `ec2.SecurityGroup()` objects, rather than some dictionary.
- To pass props to a construct and avoid the `jsii.implements` shenanigans in python, we need to add an extra step (basically what I did in python, but in typescript). See again the `Connection` class:

```ts
export class Connection extends cdk.Resource implements IConnection {
```

and how it's exported in python:

```py
@jsii.implements(IConnection)
class Connection(
    _aws_cdk_ceddda9d.Resource,
    metaclass=jsii.JSIIMeta,
    jsii_type="@aws-cdk/aws-glue-alpha.Connection",
):
```

- The same goes for the `ISecurityGroup` and `ISubnet` mentioned above (see [here](https://github.com/aws/aws-cdk/blob/4ff3565a9d7b0298bf884822fecabdd3cff643aa/packages/aws-cdk-lib/aws-ec2/lib/security-group.ts#L56), for instance)

Reply:
Yeah it is something wrong with the component, I think that [here](https://github.com/DataChefHQ/cost-monitoring-construct/blob/f5f916c28163c798d2247b2b4fea66a0c4639ea6/src/application-cost-monitoring.ts#L11) instead of 
```ts
export class ApplicationCostMonitoring extends IBudgetStrategy {
```
It should be 
```ts
export class ApplicationCostMonitoring extends Construct implements IBudgetStrategy {
```

Somewhere something goes wrong and JSII does not produce the right Python. 
![image](https://github.com/DataChefHQ/cost-monitoring-construct/assets/22810856/2814bd92-66c6-4a7c-8fd3-6287bba2e833)

VS something in the standard CDK lib, an SQS to be specific. 

![image](https://github.com/DataChefHQ/cost-monitoring-construct/assets/22810856/287ab9b7-ebeb-4a7d-9e76-5c5ac095030d)